presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -35,17 +36,20 @@
     private static final int MAX_POSITIONS_EVALUATED_PER_CALL = 10000;
 
     private final OperatorContext operatorContext;
-    private final List<Type> types;
+    private final List<Type> probeTypes;
+    private final List<Type> buildTypes;
     private final ListenableFuture<? extends LookupSource> lookupSourceFuture;
     private final JoinProbeFactory joinProbeFactory;
     private final Runnable onClose;
 
-    private final PageBuilder pageBuilder;
+    private IntVector probeIndexBuilder = new IntVector();
+    private final PageBuilder buildPageBuilder;
 
     private final boolean probeOnOuterSide;
 
     private LookupSource lookupSource;
     private JoinProbe probe;
+    private boolean reachedProbeEnd = true;
```

Comment:
Consider removing this boolean and retain signalling of probe end by `probe==null`.
Then you need to retain `probePage` somewhere (just as member here).
Then you don't need to implement `buildDictionaryPage` in probe, this can be a method here, in this class, because all it needs from probe is the probe page.